components/director/internal/domain/application/service.go

```diff
@@ -749,26 +760,100 @@ func (s *service) Merge(ctx context.Context, destID, srcID string) (*model.Appli
+	if err != nil {
+		return nil, errors.Wrapf(err, "while trying to merge labels for applications with ids %s and %s", destID, srcID)
+	}
+
 	log.C(ctx).Infof("Deleting source application with id %s", srcID)
 	if err := s.Delete(ctx, srcID); err != nil {
 		return nil, err
 	}
 
+	log.C(ctx).Infof("Deleting source labels for application with id %s", srcID)
+	if err := s.labelRepo.DeleteAll(ctx, appTenant, model.ApplicationLabelableObject, srcID); err != nil {
+		return nil, err
+	}
+
+	log.C(ctx).Infof("Updating destination app with id %s", srcID)
 	if err := s.appRepo.Update(ctx, appTenant, destApp); err != nil {
 		return nil, err
 	}
 
+	for _, destAppLabel := range destAppLabelsMerged {
+		log.C(ctx).Infof("Upserting destination label with id %s", destAppLabel.ID)
+		if err := s.labelRepo.Upsert(ctx, appTenant, destAppLabel); err != nil {
+			return nil, err
+		}
+	}
+
 	return destApp, nil
 }
 
+// handleMergeLabels merges source labels into destination labels
+func (s *service) handleMergeLabels(srcAppLabels, destAppLabels map[string]*model.Label) (map[string]*model.Label, error) {
+	destIntSysID := fmt.Sprintf("%v", destAppLabels[intSysKey].Value)
+	srcIntSysID := fmt.Sprintf("%v", srcAppLabels[intSysKey].Value)
+	if len(destIntSysID) == 0 && len(srcIntSysID) > 0 {
+		destAppLabels[intSysKey].Value = srcIntSysID
+	}
+
+	srcScenarios, ok := srcAppLabels[model.ScenariosKey].Value.([]interface{})
+	if !ok {
+		return nil, errors.Errorf("error while converting source scenario with ID %s to slice", srcAppLabels[model.ScenariosKey].ID)
+	}
+
+	destScenarios, ok := destAppLabels[model.ScenariosKey].Value.([]interface{})
+	if !ok {
+		return nil, errors.Errorf("error while converting destination scenario with ID %s to slice", destAppLabels[model.ScenariosKey].ID)
+	}
+
+	srcScenariosStrSlice := make([]string, len(srcScenarios))
+	for _, srcScenario := range srcScenarios {
+		srcScenarioStr, ok := srcScenario.(string)
+		if !ok {
+			return nil, errors.Errorf("error while converting source scenario to string")
+		}
+
+		srcScenariosStrSlice = append(srcScenariosStrSlice, srcScenarioStr)
+	}
+
+	destScenariosStrSlice := make([]string, 0, len(destScenarios))
+	for _, destScenario := range destScenarios {
+		destScenarioStr, ok := destScenario.(string)
+		if !ok {
+			return nil, errors.Errorf("error while converting destination scenario to string")
+		}
+
+		destScenariosStrSlice = append(destScenariosStrSlice, destScenarioStr)
+	}
+
+	for _, srcScenario := range srcScenariosStrSlice {
+		if len(srcScenario) > 0 && !containsStr(destScenariosStrSlice, srcScenario) {
+			destScenariosStrSlice = append(destScenariosStrSlice, srcScenario)
+		}
+	}
+
+	if err := mergo.Merge(&destAppLabels, srcAppLabels); err != nil {
+		return nil, errors.Wrapf(err, "while trying to merge labels")
+	}
+
+	destAppLabels[model.ScenariosKey].Value = destScenariosStrSlice
+	destAppLabels[managedKey].Value = srcAppLabels[managedKey].Value
+
```

Comment:
I was wondering why some of the labels have to be manually assigned instead of just leveraging the mergo lib..?

Maybe there's some limitations to the lib - if so maybe it's worth noting in the function go doc?

Reply:
Limitations, yeah. That's why I need to handle the scenarios and intSysID manually. The case for IntSysID is that if any of the IDs (source or dest) is empty string, it ignores the whole field and does not do the merging. You can specify options for the merging there was one "Overwrite" which would do the trick in the case, but it would merge all fields from src in dest even if the src already has them populated. Which is not what we want. Will add more info in the godoc